game/engine.py

```
@@ -99,11 +157,122 @@ def from_dict(cls, data):
+            game.repetition_history = [game.generate_position_key()]
+
+        game._rebuild_repetition_counts()
+
+        game.valid_moves_cache = {}
+        return game
 
+    @classmethod
+    def from_fen(cls, fen: str, time_limit=600):
+        """Create a new game state from a FEN string (board, side, castling)."""
+        if not isinstance(fen, str):
+            raise ValueError("FEN must be a string.")
```

Ruff (0.15.14)

[warning] Prefer `TypeError` exception for invalid type

(TRY004)